File: gitjuggling/src/prune.rs
use crate::discover::LocalRepos;
use crate::remote::RemoteRepo;
use std::path::PathBuf;

/// A local repo that has no matching upstream remote.
#[derive(Debug)]
pub struct OrphanRepo {
    pub path: PathBuf,
    pub name: String,
}

/// Find local repos that have no matching upstream remote.
pub fn find_orphans(local: &LocalRepos, remote_repos: &[RemoteRepo]) -> Vec<OrphanRepo> {
    // Build a set of normalized remote URLs for fast lookup
    let remote_urls: std::collections::HashSet<String> = remote_repos
        .iter()
        .map(|r| normalize_for_lookup(&r.clone_url))
        .collect();

    let mut orphans = Vec::new();

    for repo in &local.repos {
        // Check if any of the remote URLs match an upstream
        let has_match = repo
            .remote_urls
            .values()
            .any(|url| remote_urls.contains(&normalize_for_lookup(url)));

        if !has_match {
            let name = repo
                .path
                .file_name()
                .map(|n| n.to_string_lossy().to_string())
                .unwrap_or_default();

            orphans.push(OrphanRepo {
                path: repo.path.clone(),
                name,
            });
        }
    }

    orphans
}

/// Result of pruning a single orphan.
#[derive(Debug)]
pub struct PruneResult {
    pub name: String,
    #[allow(dead_code)]
    pub path: PathBuf,
    pub success: bool,
    pub message: String,
}

/// Prune orphan repos (interactive or dry-run).
pub fn prune_orphans(orphans: &[OrphanRepo], dry_run: bool, interactive: bool) -> Vec<PruneResult> {
    let mut results = Vec::new();

    for orphan in orphans {
        if dry_run {
            results.push(PruneResult {
                name: orphan.name.clone(),
                path: orphan.path.clone(),
                success: true,
                message: "would remove".to_string(),
            });
            continue;
        }

        if interactive {
            let prompt = format!(
                "Remove orphan repo {} ({})?",
                orphan.name,
                orphan.path.display()
            );
            let confirm = dialoguer::Confirm::new()
                .with_prompt(&prompt)
                .default(false)
                .interact();

            match confirm {
                Ok(false) => {
                    results.push(PruneResult {
                        name: orphan.name.clone(),
                        path: orphan.path.clone(),
                        success: true,
                        message: "skipped (user declined)".to_string(),
                    });
                    continue;
                }
                Err(err) => {
                    results.push(PruneResult {
                        name: orphan.name.clone(),
                        path: orphan.path.clone(),
                        success: false,
                        message: format!("prompt failed: {}", err),
                    });
                    continue;
                }
                Ok(true) => {}
            }
        }

        match std::fs::remove_dir_all(&orphan.path) {
            Ok(()) => results.push(PruneResult {
                name: orphan.name.clone(),
                path: orphan.path.clone(),
                success: true,
                message: "removed".to_string(),
            }),
            Err(err) => results.push(PruneResult {
                name: orphan.name.clone(),
                path: orphan.path.clone(),
                success: false,
                message: format!("failed to remove: {}", err),
            }),
        }
    }

    results
}

/// Normalize a URL for lookup comparison.
/// Reuses the same logic as discover::normalize_url.
fn normalize_for_lookup(url: &str) -> String {
    let url = url.trim().trim_end_matches('/').trim_end_matches(".git");

    // SSH form: git@host:owner/repo
    if let Some(rest) = url.strip_prefix("git@") {
        return rest.replacen(':', "/", 1);
    }

    // HTTPS form
    if let Some(rest) = url
        .strip_prefix("https://")
        .or_else(|| url.strip_prefix("http://"))
    {
        return rest.to_string();
    }

    url.to_string()
}

/// Print a summary of prune results.
pub fn print_prune_summary(results: &[PruneResult]) {
    if results.is_empty() {
        println!("\n  {}", "No orphan repos found.".dimmed());
        return;
    }

    let (removed, failed): (Vec<_>, Vec<_>) = results.iter().partition(|r| r.success);

    use colored::Colorize;
    println!(
        "\n{}{}{}",
        "=== ".bright_white(),
        "Prune Summary".bright_cyan(),
        " ===".bright_white()
    );

    for r in &removed {
        println!("  {} {} ({})", "✓".green(), r.name, r.message);
    }

    for r in &failed {
        println!("  {} {} — {}", "✗".red(), r.name, r.message.red());
    }

    println!(
        "\n{} {} | {} {}",
        "Removed:".blue(),
        removed.len().to_string().bright_green(),
        "Failed:".blue(),
        failed.len().to_string().bright_red(),
    );
}

File: gitjuggling/internal/prune/prune.go
package prune

import (
	"fmt"
	"os"
	"path/filepath"

	"dev.rischmann.fr/mytools/gitjuggling/internal/discover"
	"dev.rischmann.fr/mytools/gitjuggling/internal/remote"
)

// OrphanRepo represents a local repo that has no matching upstream remote.
type OrphanRepo struct {
	Path string
	Name string
}

// PruneResult holds the outcome of pruning a single orphan repo.
type PruneResult struct {
	Name    string
	Path    string
	Success bool
	Message string
}

// ConfirmFunc is a callback for interactive confirmation prompts.
type ConfirmFunc func(prompt string) bool

// FindOrphans finds local repos that have no matching upstream remote.
func FindOrphans(local *discover.LocalRepos, remoteRepos []*remote.RemoteRepo) []*OrphanRepo {
	// Build a set of normalized remote URLs for fast lookup
	remoteURLs := make(map[string]struct{}, len(remoteRepos))
	for _, r := range remoteRepos {
		remoteURLs[discover.NormalizeURL(r.CloneURL)] = struct{}{}
	}

	var orphans []*OrphanRepo

	for _, repo := range local.Repos {
		hasMatch := false
		for _, url := range repo.RemoteURLs {
			if _, ok := remoteURLs[discover.NormalizeURL(url)]; ok {
				hasMatch = true
				break
			}
		}

		if !hasMatch {
			name := filepath.Base(repo.Path)
			orphans = append(orphans, &OrphanRepo{
				Path: repo.Path,
				Name: name,
			})
		}
	}

	return orphans
}

// PruneOrphans removes orphan repos. If dryRun is true, it only reports what
// would be done. If confirmFn is non-nil, it prompts for each orphan.
func PruneOrphans(orphans []*OrphanRepo, dryRun bool, confirmFn ConfirmFunc) []*PruneResult {
	var results []*PruneResult

	for _, orphan := range orphans {
		if dryRun {
			results = append(results, &PruneResult{
				Name:    orphan.Name,
				Path:    orphan.Path,
				Success: true,
				Message: "would remove",
			})
			continue
		}

		if confirmFn != nil {
			prompt := fmt.Sprintf("Remove orphan repo %s (%s)?", orphan.Name, orphan.Path)
			if !confirmFn(prompt) {
				results = append(results, &PruneResult{
					Name:    orphan.Name,
					Path:    orphan.Path,
					Success: true,
					Message: "skipped (user declined)",
				})
				continue
			}
		}

		if err := os.RemoveAll(orphan.Path); err != nil {
			results = append(results, &PruneResult{
				Name:    orphan.Name,
				Path:    orphan.Path,
				Success: false,
				Message: fmt.Sprintf("failed to remove: %v", err),
			})
		} else {
			results = append(results, &PruneResult{
				Name:    orphan.Name,
				Path:    orphan.Path,
				Success: true,
				Message: "removed",
			})
		}
	}

	return results
}
